### PR TITLE
feat: add verify command

### LINKED ISSUE #2645: Verify the .golangci.yaml config
### Your feature request related to a problem? Please describe.

Crafting the .golangci.yaml can be a little challenging as the config grows or just getting right the first time. For example, I was crafting a config and had my indention off for the `linters-settings:` and rather than error'ing, the linter-settings were just not applied. 

### Describe the solution you'd like.

I propose a `verify` command is added to the `golangci-lint` CLI tool that will verify the config schema. 

### Describe alternatives you've considered.

Alternatively, `golangci-lint run` could throw an error when the config does not conform to the top-level schema. 

### Additional context.

_No response_

### PR DESCRIPTION
**Pull request by ldez**
This PR adds a new command `config verify`.

This command allows to verify a configuration file against a JSON Schema.

The JSON schema is obtained based on this approach:
- there is a version -> it uses `https://golangci-lint.run/jsonschema/golangci.v<major>.<minor>.jsonschema.json`
- there is a commit hash/revision -> it uses `https://raw.githubusercontent.com/golangci/golangci-lint/<commit>/jsonschema/golangci.next.jsonschema.json`
- there is no valid information -> error

The make target `website_copy_jsonschema` has been modified to create `https://golangci-lint.run/jsonschema/golangci.v<major>.<minor>.jsonschema.json`.

Fixes #2645

---

**Why we will not use JSON Schema to validate `.golangci.reference.yml` or `.golangci.next.reference.yml`?**

`.golangci.reference.yml` and `.golangci.next.reference.yml` are invalid by nature: they contain the exhaustive list of options but those options can be incompatible.

```console
$ go run ./cmd/golangci-lint config verify -c .golangci.next.reference.yml --schema ./jsonschema/golangci.next.jsonschema.json
jsonschema: linters-settings.goheader does not validate with /properties/linters-settings/properties/goheader/oneOf: valid against schemas at indexes 0 and 1
jsonschema: linters-settings.custom.example does not validate with /properties/linters-settings/properties/custom/patternProperties/%5E.%2A$/oneOf: valid against schemas at indexes 0 and 1
```
Same thing for the configuration loading:
```console
$ go run ./cmd/golangci-lint run -c .golangci.next.reference.yml                                       
ERRO [linter] thelper: at least one option must be enabled 
exit status 3
```

**How many versions of the JSON schema we will support?**

The build of `golangci-lint` using a specific official commit will use the schema related to its commit (`https://raw.githubusercontent.com/golangci/golangci-lint/<commit>/jsonschema/golangci.next.jsonschema.json`).

For the "real" versions, for now, we will support only the current stable version.

The system has been made to support more versions if needed: if a schema is inside the directory `jsonschema/` it will be published as the other schema.
If we want to follow this way, I think we should limit the support to the 2 latest stable versions.

**Why we don't embed the JSON Schema?**

The usage of remote JSON schema allows us to fix the schema without releasing golangci-lint.

I think IDEs or text editors, that use JSON schema, will be always better than a command line to quickly identify syntax errors.
It is not worth to increase the binary size only for a niche.

### CODE REVIEW

**Comment by ldez** [comment]
@Antonboom do you have an opinion about this one?

File: pkg/commands/config.go [hunk 1]
```diff
@@ -33,6 +44,16 @@ func newConfigCommand(log logutils.Log) *configCommand {
 		RunE: func(cmd *cobra.Command, _ []string) error {
 			return cmd.Help()
 		},
+		PersistentPreRunE: c.preRunE,
+	}
+
+	verifyCommand := &cobra.Command{
```
Comment (alexandear) [inline pkg/commands/config.go:50]: Could we enhance the help message for the config verify command to display all possible flags?

Currently, it's not clear what should be provided:
```
❯ ./golangci-lint config verify -h
Verify configuration against JSON schema

Usage:
  golangci-lint config verify [flags]

Global Flags:
      --color string   Use color when printing; can be 'always', 'auto', or 'never' (default "auto")
  -c, --config PATH    Read config from file path PATH
  -h, --help           Help for a command
      --no-config      Don't read config file
  -v, --verbose        Verbose output
```

Yes, someone might look at this PR and notice the `--schema` parameter, but that's not a typical user.

Comment (alexandear) [reply pkg/commands/config.go:50]: The help might look like this:

```
❯ ./golangci-lint config verify -h
Verify configuration against JSON schema

Usage:
  golangci-lint config verify [flags]

Flags:
  -c, --config PATH   Read config from file path PATH
      --no-config     Don't read config file
  -s, --schema        JSON schema path

Global Flags:
      --color string   Use color when printing; can be 'always', 'auto', or 'never' (default "auto")
  -h, --help           Help for a command
  -v, --verbose        Verbose output
```

Comment (ldez) [reply pkg/commands/config.go:50]: The flag is only for debugging purposes, it's not a user flag.

Comment (ldez) [reply pkg/commands/config.go:50]: It's already consistent:

````console
 $ ./golangci-lint config path -h                                            
Print used config path

Usage:
  golangci-lint config path [flags]

Global Flags:
      --color string   Use color when printing; can be 'always', 'auto', or 'never' (default "auto")
  -c, --config PATH    Read config from file path PATH
  -h, --help           Help for a command
      --no-config      Don't read config file
  -v, --verbose        Verbose output
```

File: pkg/commands/config_verify.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	err = validateConfiguration(schemaURL, usedConfigFile)
+	if err != nil {
+		var v *jsonschema.ValidationError
+		if !errors.As(err, &v) {
+			return fmt.Errorf("[%s] validate: %w", usedConfigFile, err)
+		}
+
+		detail := v.DetailedOutput()
+
+		printValidationDetail(cmd, &detail)
```
Comment (alexandear) [inline pkg/commands/config_verify.go:42]: Do we need to return a non-zero exit code when config validation fails?

File: pkg/commands/config_verify.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	}
+
+	return nil
+}
+
+func printValidationDetail(cmd *cobra.Command, detail *jsonschema.Detailed) {
+	if detail.Error != "" {
+		cmd.PrintErrf("jsonschema: %s does not validate with %s: %s\n",
```
Comment (alexandear) [inline pkg/commands/config_verify.go:50]: The current golangci-lint output when the config is invalid:
```
❯ ./golangci-lint config verify --schema jsonschema/golangci.next.jsonschema.json
jsonschema: issues does not validate with /properties/issues/additionalProperties: additionalProperties 'exclude-paths' not allowed
jsonschema: run does not validate with /properties/run/additionalProperties: additionalProperties 'execute-timeout', 'skip-dirs' not allowed
```

Maybe we should improve it:
```
❯ ./golangci-lint config verify --schema jsonschema/golangci.next.jsonschema.json
The config is invalid:
jsonschema: "issues" does not validate with "/properties/issues/additionalProperties": additionalProperties 'exclude-paths' not allowed
jsonschema: "run" does not validate with "/properties/run/additionalProperties": additionalProperties 'execute-timeout', 'skip-dirs' not allowed
```
?

File: pkg/commands/config_verify.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+			strings.ReplaceAll(strings.TrimPrefix(detail.InstanceLocation, "/"), "/", "."), detail.KeywordLocation, detail.Error)
+	}
+
+	for _, d := range detail.Errors {
+		d := d
+		printValidationDetail(cmd, &d)
+	}
+}
+
+func getSchemaURL(flags *pflag.FlagSet, buildInfo BuildInfo) (string, error) {
```
Comment (alexandear) [inline pkg/commands/config_verify.go:60]: `getSchemaURL` violates [SRP](https://en.wikipedia.org/wiki/Single_responsibility_principle) by performing two independent tasks:
- retrieving `schemaURL` from the flag set;
- retrieving `schemaURL` from the build info.

We might want to consider refactoring it.

File: pkg/commands/config_verify.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	defer func() { _ = file.Close() }()
+
+	var m any
+	err = toml.NewDecoder(file).Decode(&m)
+	if err != nil {
+		return nil, fmt.Errorf("[%s] TOML decode: %w", filename, err)
+	}
+
+	return m, nil
+}
```
Comment (alexandear) [inline pkg/commands/config_verify.go:175]: These functions have the same logic. We can think of refactoring them.

File: pkg/commands/config_verify.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	"github.com/santhosh-tekuri/jsonschema/v5"
+	_ "github.com/santhosh-tekuri/jsonschema/v5/httploader"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"gopkg.in/yaml.v3"
+
+	"github.com/golangci/golangci-lint/pkg/exitcodes"
+)
+
+func (c *configCommand) executeVerify(cmd *cobra.Command, _ []string) error {
```
Comment (alexandear) [inline pkg/commands/config_verify.go:21]: Please add unit tests that cover at least successful cases.

File: pkg/commands/config_verify.go (hunk 4 above)
Comment (ldez) [reply pkg/commands/config_verify.go:60]: I disagree with that, I will just change the name of the function.

File: pkg/commands/config_verify.go (hunk 5 above)
Comment (ldez) [reply pkg/commands/config_verify.go:175]: I think it's over-factorization/technical-factorization: the 2 functions are close but they are readable and dedicated to a specific need.
To factorize we will have to create an interface and in the end, we will have a less readable technical "one-liner".

File: pkg/commands/config_verify.go (hunk 6 above)
Comment (ldez) [reply pkg/commands/config_verify.go:21]: There is a test inside the CI: step `commands`

File: pkg/commands/config_verify.go (hunk 4 above)
Comment (alexandear) [reply pkg/commands/config_verify.go:60]: Technically, "renaming a function" is "refactoring". I agree with this change.

File: pkg/commands/config_verify.go (hunk 6 above)
Comment (ldez) [reply pkg/commands/config_verify.go:25]: And it works well :smile: https://github.com/golangci/golangci-lint/actions/runs/8346859417/job/22845173325

I updated the schema used by this file.

I also added tests on `createSchemaURL`.

**Review by alexandear — APPROVED** [review]

File: scripts/website/copy_jsonschema/main.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		log.Fatalf("get latest release version: %v", err)
+	}
+
+	version, err := hcversion.NewVersion(latest)
+	if err != nil {
+		log.Fatalf("parse version: %v", err)
+	}
+
+	versioned := fmt.Sprintf("golangci.v%d.%d.jsonschema.json", version.Segments()[0], version.Segments()[1])
+	files[versioned] = "golangci.jsonschema.json"
```
Comment (Antonboom) [inline scripts/website/copy_jsonschema/main.go:55]: will `verify` support custom gcl schema in the future?

**Review by Antonboom — APPROVED** [review]
> @Antonboom do you have an opinion about this one?

🙏 I try to review as soon as possible

Let's consider the feature with validator in golangci website, like

- https://envoylint.com/
- https://docs.gitlab.com/ee/ci/lint.html
- https://config.travis-ci.com/explore
- ...

File: scripts/website/gh/gh.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return "", fmt.Errorf("get HTTP response for the latest tag: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
```
[the remark this answers is not in the document]
Comment (ldez) [reply scripts/website/gh/gh.go:37]: https://github.com/golangci/golangci-lint/pull/4502#discussion_r1524905012

File: scripts/website/gh/gh.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
+package gh
```
[the remark this answers is not in the document]
Comment (ldez) [reply scripts/website/gh/gh.go:1]: I will never create something called `utils` or `helper`, this is the devil!

But I agree the current name can be unclear.

File: scripts/website/copy_jsonschema/main.go (hunk 7 above)
Comment (ldez) [reply scripts/website/copy_jsonschema/main.go:55]: Maybe as a dedicated command under `custom` :thinking: 
`config` is a command related to `golangci-lint` configuration, then `verify` too.

File: pkg/commands/custom.go [hunk 10]
```diff
@@ -25,11 +25,12 @@ func newCustomCommand(logger logutils.Log) *customCommand {
...
-		Short:   "Build a version of golangci-lint with custom linters",
-		Args:    cobra.NoArgs,
-		PreRunE: c.preRunE,
-		RunE:    c.runE,
+		Use:          "custom",
+		Short:        "Build a version of golangci-lint with custom linters",
+		Args:         cobra.NoArgs,
+		PreRunE:      c.preRunE,
+		RunE:         c.runE,
+		SilenceUsage: true,
```
[the remark this answers is not in the document]
Comment (ldez) [reply pkg/commands/custom.go:33]: yes and no: I added on the `verify` and then I propagated the change.

**Comment by ldez** [comment]
> Let's consider the feature with validator in golangci website

Before doing that the website should be changed.
I'm not sure about the additional value for golangci-lint: the links you provided are mainly related to tools that users don't have on their computers. golangci-lint is installed and runs on computers.

File: pkg/commands/custom.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ldez) [reply pkg/commands/custom.go:33]: I reverted the commit.

**Pull request merged**